gpt4all-backend/replit.cpp

```diff
@@ -0,0 +1,1131 @@
+            const int32_t erasePoint = promptCtx.n_ctx * promptCtx.contextErase;
+            // Erase the first percentage of context from the tokens...
+            std::cerr << "Replit: reached the end of the context window so resizing\n";
+            promptCtx.tokens.erase(promptCtx.tokens.begin(), promptCtx.tokens.begin() + erasePoint);
+            promptCtx.n_past = promptCtx.tokens.size();
+            recalculateContext(promptCtx, recalculateCallback);
+            assert(promptCtx.n_past + 1 <= promptCtx.n_ctx);
+        }
+
+        const int64_t t_start_predict_us = ggml_time_us();
+        if (!replit_eval(*d_ptr->model, d_ptr->n_threads, promptCtx.n_past, { id }, promptCtx.logits,
+            d_ptr->mem_per_token)) {
+            std::cerr << "Replit ERROR: Failed to predict next token\n";
+            return;
+        }
+        t_predict_us += ggml_time_us() - t_start_predict_us;
+
+        promptCtx.n_past += 1;
+        // display text
+        ++totalPredictions;
+
+        // mpt-7b-chat has special token for end
+        // TODO: need to change this for replit model (return, ```, def)
+        if (d_ptr->has_end_of_text && id == d_ptr->vocab.raw_vocab.token_to_id["<|endoftext|>"])
+            goto stop_generating;
+
+        if (id == 0 ) // end of text
+            goto stop_generating;
+
+        //const std::string str = d_ptr->vocab.raw_vocab.id_to_token[id];
+        const std::string str = replit_tokenizer_detokenize(d_ptr->vocab, {static_cast<std::size_t>(id)}).c_str();
+
+        // Check if the provided str is part of our reverse prompts
+        bool foundPartialReversePrompt = false;
+        const std::string completed = cachedResponse + str;
+        if (reversePrompts.find(completed) != reversePrompts.end()) {
+            goto stop_generating;
+        }
+
+        // Check if it partially matches our reverse prompts and if so, cache
+        for (auto s : reversePrompts) {
+            if (s.compare(0, completed.size(), completed) == 0) {
+                foundPartialReversePrompt = true;
+                cachedResponse = completed;
+                break;
+            }
+        }
+
+        // Regardless the token gets added to our cache
+        cachedTokens.push_back(id);
+
+        // Continue if we have found a partial match
+        if (foundPartialReversePrompt)
+            continue;
+
+        // Empty the cache
+        for (auto t : cachedTokens) {
+            if (promptCtx.tokens.size() == promptCtx.n_ctx)
+                promptCtx.tokens.erase(promptCtx.tokens.begin());
+            promptCtx.tokens.push_back(t);
+            //printf("%s", replit_tokenizer_detokenize(d_ptr->vocab, {static_cast<std::size_t>(t)}).c_str());
+            if (!responseCallback(t, replit_tokenizer_detokenize(d_ptr->vocab, {static_cast<std::size_t>(t)}).c_str()))
+                goto stop_generating;
+        }
+        cachedTokens.clear();
+    }
+
+stop_generating:
+
+#if 0
```

Comment:
If this is dead code then also remove all the now useless timestamp variables and calls.

Reply:
This is from previous code in ggml examples. He's just copying what he found there.